mypy/checker.py

```diff
@@ -1618,19 +1628,21 @@ def check_compatibility(self, name: str, base1: TypeInfo,
             return
         first = base1[name]
         second = base2[name]
-        first_type = first.type
-        if first_type is None and isinstance(first.node, FuncBase):
-            first_type = self.function_type(first.node)
-        second_type = second.type
-        if second_type is None and isinstance(second.node, FuncBase):
-            second_type = self.function_type(second.node)
+        first_type = self.determine_type_of_class_member(first)
+        second_type = self.determine_type_of_class_member(second)
+
         # TODO: What if some classes are generic?
         if (isinstance(first_type, FunctionLike) and
                 isinstance(second_type, FunctionLike)):
-            # Method override
-            first_sig = bind_self(first_type)
-            second_sig = bind_self(second_type)
-            ok = is_subtype(first_sig, second_sig, ignore_pos_arg_names=True)
+            if first_type.is_type_obj() and second_type.is_type_obj():
+                # For class objects only check the subtype relationship of the classes,
+                # since we allow incompatible overrides of '__init__'/'__new__'
+                ok = is_subtype(left=fill_typevars(first_type.type_object()),
+                                right=fill_typevars(second_type.type_object()))
```

Comment:
I think using `fill_typevars` here is dangerous, since it may produce some false positives for generic classes (it creates an instance with type variable types). You need to apply `erase_typevars`. Or even better, add a helper function to `typevars.py` that will be very similar to `fill_typevars`, but would insert correct number of `Any` as type arguments.